File: frontend/src/types/websocket.ts
export interface WebSocketMessage {
  type: 'step_update' | 'search_complete' | 'search_cancelled' | 'error' | 'session_reset';
  search_id?: string;
  data: any;
  timestamp: string;
}

export interface StepUpdateMessage extends WebSocketMessage {
  type: 'step_update';
  data: {
    step_id: string;
    step_type: 'plan' | 'search' | 'code' | 'solve' | 'replan';
    status: 'started' | 'running' | 'completed' | 'failed';
    title: string;
    content?: string;
    metadata?: {
      searchQuery?: string;
      codeResult?: string;
      planSteps?: string[];
      error?: string;
    };
  };
}

export interface SearchCompleteMessage extends WebSocketMessage {
  type: 'search_complete';
  data: {
    search_id: string;
    result: string;
    total_steps: number;
    duration: number;
  };
}

export interface ErrorMessage extends WebSocketMessage {
  type: 'error';
  data: {
    error: string;
    step_id?: string;
    recoverable: boolean;
  };
}

export interface SearchCancelledMessage extends WebSocketMessage {
  type: 'search_cancelled';
  data: {
    search_id: string;
    message: string;
  };
}

export interface SessionResetMessage extends WebSocketMessage {
  type: 'session_reset';
  data: {
    message: string;
  };
}

File: backend/app/models/websocket.py
from datetime import datetime
from typing import Optional, Dict, Any, Literal, Union
from pydantic import BaseModel, Field
from .search import StepType, StepStatus, StepMetadata

class WebSocketMessage(BaseModel):
    type: str
    timestamp: datetime
    data: Dict[str, Any]
    search_id: Optional[str] = None

class StepUpdateData(BaseModel):
    step_id: str
    step_type: StepType
    status: StepStatus
    title: str
    content: Optional[str] = ""
    metadata: Optional[StepMetadata] = None

class StepUpdateMessage(WebSocketMessage):
    type: Literal["step_update"] = "step_update"
    data: StepUpdateData

class SearchCompleteData(BaseModel):
    search_id: str
    result: str
    total_steps: int
    duration: float
    final_answer: str

class SearchCompleteMessage(WebSocketMessage):
    type: Literal["search_complete"] = "search_complete"
    data: SearchCompleteData

class ErrorData(BaseModel):
    error: str
    step_id: Optional[str] = None
    recoverable: bool = True
    error_code: Optional[str] = None

class ErrorMessage(WebSocketMessage):
    type: Literal["error"] = "error"
    data: ErrorData

class SessionResetData(BaseModel):
    message: str = "Session has been reset"
    reason: Optional[str] = None

class SessionResetMessage(WebSocketMessage):
    type: Literal["session_reset"] = "session_reset"
    data: SessionResetData

class ConnectionData(BaseModel):
    connected: bool
    client_id: str
    server_time: datetime

class ConnectionMessage(WebSocketMessage):
    type: Literal["connection"] = "connection"
    data: ConnectionData

class HeartbeatData(BaseModel):
    server_time: datetime
    client_count: int

class HeartbeatMessage(WebSocketMessage):
    type: Literal["heartbeat"] = "heartbeat"
    data: HeartbeatData

# Union type for all possible WebSocket messages
WSMessage = Union[
    StepUpdateMessage,
    SearchCompleteMessage,
    ErrorMessage,
    SessionResetMessage,
    ConnectionMessage,
    HeartbeatMessage
]
